sharding/collator.go

```diff
@@ -0,0 +1,106 @@
+package sharding
+
+import (
+	"context"
+	"fmt"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"math/big"
+)
+
+// subscribeBlockHeaders checks incoming block headers and determines if
+// we are an eligible proposer for collations. Then, it finds the pending tx's
+// from the running geth node and sorts them by descending order of gas price,
+// eliminates those that ask for too much gas, and routes them over
+// to the VMC to create a collation
+func subscribeBlockHeaders(c *Client) error {
+	headerChan := make(chan *types.Header, 16)
+
+	_, err := c.client.SubscribeNewHead(context.Background(), headerChan)
```

Comment:
Please add a TODO for handling error conditions for this subscription, i.e. getting disconnected.